[eunrcn] iP

eunrcn opened the pull request:
# *Skyler Feature Enhancement*

## *Changes Made*

1. Implemented new features and improvements for Skyler.
2. Enhanced user experience with text-based interactions.
3. Added functionality to manage tasks efficiently.

## *Features Added*

- Managing tasks with Skyler is now text-based and easy to learn.
- Skyler is FAST, SUPER FAST to use.
- Download Skyler [here](https://github.com/eunrcn/ip).
- Double-click the downloaded file to launch Skyler.
- Add your tasks and let Skyler manage them for you 😉.
- Skyler is FREE!

## *Features*

[x] Managing tasks.
[x] Managing deadlines.
[   ] Reminders.

> skyler is the name of a very cutie doge

## *For Java Programmers*

If you are a Java programmer, you can use DukePro to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

eunrcn committed:
Base Version: Level 0 - Rename, Greet, Exit

eunrcn committed:
Level 1: Echo

eunrcn committed:
Level 2: Add, List

eunrcn committed:
Level 3. Mark as Done

eunrcn committed:
Level 4. ToDos, Events, Deadlines

eunrcn committed:
A-TextUiTesting: Automated TextD UI Testing

eunrcn committed:
Level 5. Handle Errors

eunrcn committed:
Level 6. Delete

eunrcn committed:
branch-level-7

eunrcn committed:
branch-level-7

eunrcn committed:
Merge branch 'branch-Level-7'

eunrcn committed:
branch-level-8

eunrcn committed:
Teaching the chatbot to understand dates and times

Dates and times are crucial for accurate task management. Implementing Level 8 by storing deadline dates as java.time.LocalDate objects and accepting dates in yyyy-mm-dd format. Print dates in a more readable format (MMM dd yyyy). Additionally, added a command to view deadlines/events on a specific date. This improves the overall functionality and user experience of the chatbot.

Level 8. Dates and Times
- Implemented storage of deadline dates as java.time.LocalDate.
- Accepted dates in yyyy-mm-dd format and printed in MMM dd yyyy format.
- Added a command to view deadlines/events on a specific date.

eunrcn committed:
Merge branch 'branch-Level-8'

eunrcn committed:
phew

eunrcn committed:
Refactor the code to extract out closely related code as classes

Extract the following classes:
- Ui: Handles user interactions
- Storage: Manages loading and saving tasks from/to the file
- Parser: Processes user commands
- TaskList: Manages the task list with add/delete operations

The code needs restructuring to improve modularity and readability. Each class now has a clear responsibility, making the code more maintainable.

Implemented the extraction of related code into separate classes to enhance code organization and maintainability.

Each class is designed to handle a specific aspect of the application, improving code cohesion and clarity.

The extraction is done to follow a more object-oriented design, promoting better code organization and ease of maintenance.

Additional info: The class names and responsibilities align with standard design principles, enhancing code understanding for future development.

eunrcn committed:
Merge A-MoreOOP

eunrcn committed:
Added Packages

eunrcn committed:
Added Git.txt

eunrcn committed:
Merge A-Packages

eunrcn committed:
Merge add-gradle-support

eunrcn committed:
Added JUnit tests that can be run with ./gradlew test

eunrcn committed:
Added JavaDocs

eunrcn committed:
Added Find method

SimHanKiong commented on src/main/java/skyler/main/Skyler.java:
```diff
@@ -0,0 +1,34 @@
+import skyler.exception.SkylerException;
+
+public class Skyler {
+    private static final String CHATBOT_NAME = "Skyler";
```
Nice job for correct constant syntax

shaunlxw replied:
I agree!

SimHanKiong commented on src/main/java/skyler/main/Skyler.java:
```diff
@@ -0,0 +1,34 @@
+        Ui.printGreeting(CHATBOT_NAME);
+
+        Storage.loadTasksFromFile();
+        Scanner scanner = new Scanner(System.in);
+
+        while (true) {
+            String userInput = Ui.getUserInput(scanner);
+
+            if (userInput.equals("bye")) {
+                Ui.printByeMessage();
+                scanner.close();
+                Storage.saveTasksToFile(); // Save tasks before exiting
+                break;
```
Would putting the scanner in the Ui class be better?

SimHanKiong commented on src/main/java/skyler/main/TaskList.java:
```diff
@@ -0,0 +1,192 @@
+            if (isValidTaskId(taskId)) {
+                Task task = tasks.get(taskId - 1);
+                task.markAsUndone();
+                System.out.println("Skyler: OK, I've marked this task as not done yet:");
+                System.out.println("  " + task);
+                System.out.println("------------------------------------------------------------");
+                Storage.saveTasksToFile();
```
What do you think about moving this to the Ui class?

SimHanKiong commented on src/main/java/skyler/main/TaskList.java:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        if (matchingTasks.isEmpty()) {
+            System.out.println("Skyler: No matching tasks found.");
+        } else {
+            System.out.println("Skyler: Here are the matching tasks in your list:");
+            for (int i = 0; i < matchingTasks.size(); i++) {
+                System.out.println(" " + (i + 1) + "." + matchingTasks.get(i));
+            }
+        }
+
+        System.out.println("------------------------------------------------------------");
```
Maybe can consider passing TaskList to Ui class and print from there?

SimHanKiong reviewed (commented):
LGTM. Very clean code. Was a pleasure reviewing through it. Left a few minor suggestions to consider if you would like to.

eunrcn committed:
CheckStyle Integration
↳ Detect and Enforce Coding Style

Implementing CheckStyle to identify and address coding style violations..

eunrcn committed:
Merge A-checkstyle

shaunlxw commented on build.gradle:
```diff
@@ -0,0 +1,53 @@
+    mainClass.set("skyler.main.Skyler")
+}
+
+shadowJar {
+    archiveFileName = 'skyler.jar'
+}
+
+shadowJar {
+    archiveBaseName = "duke"
+    archiveClassifier = null
+    dependsOn("distZip", "distTar")
+}
```
I think the shadowJar properties can be put into a single shadowJar object

shaunlxw commented on src/main/java/skyler/main/Storage.java:
```diff
@@ -0,0 +1,107 @@
+        String details = data.substring(7).trim(); // Extracting task details
+
+        switch (taskType) {
+            case "T":
```
Note the switch statement coding style defined in the [given Java coding standard](https://se-education.org/guides/conventions/java/intermediate.html#layout). The case and default clauses shouldn't be indented. You can configure IntelliJ or your IDE to handle this automatically.

shaunlxw commented on src/main/java/skyler/task/Event.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + from.format(DateTimeFormatter.ofPattern("MMM dd yyyy")) +
```
Line breaks should be before an operator. The + should be on the next line.

shaunlxw commented on src/test/java/skyler/main/ParserTest.java:
```diff
@@ -0,0 +1,24 @@
+public class ParserTest {
+
+    @Test
+    public void processUserInput_eventCommand_addsEventTask() throws SkylerException {
```
Expected behaviour in the naming of the test method could be in past tense? I.e. processUserInput_eventCommand_eventTaskAdded. I am not sure about this but the examples shown are given as such.

shaunlxw commented on src/test/java/skyler/main/ParserTest.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    @Test
+    public void processUserInput_invalidCommand_throwsSkylerException() {
```
```suggestion
    public void processUserInput_invalidCommand_skylerExceptionThrown() {
```

shaunlxw reviewed (commented):
Follows coding standard very well! Only issue spotted is the indentation for the case and default clause in the switch statement. 👍

eunrcn committed:
Implement Level 10: GUI

Integrate JavaFX technology to create a graphical user interface for the chatbot. Following the SE-EDU/guides JavaFX tutorial

eunrcn committed:
Merge branch-level-10

eunrcn committed:
{Current Situation}
The parseTaskFromFile method does not include assertions to check for a non-null data parameter.

{Why It Needs to Change}
To enhance code robustness and catch potential issues early, it is crucial to ensure that method parameters are not null, especially in critical parsing operations.

{What Is Being Done About It}
Assertions have been added to the parseTaskFromFile method to verify that the data parameter is not null before proceeding with task parsing.

{Why It Is Done That Way}
By employing assertions, we establish a defensive measure during development, preventing potential null pointer issues in the parseTaskFromFile method. This ensures a more robust and reliable parsing process.

{Any Other Relevant Info}
Assertions contribute to code quality by expressing and validating assumptions, providing a clear mechanism for documenting and enforcing critical conditions in the code. Remember to enable assertions during development for them to take effect.

eunrcn committed:
Merge Assertion Branch

eunrcn committed:
Update README.md

eunrcn committed:
Update README.md

eunrcn committed:
Update README.md

eunrcn committed:
Create main.yml

eunrcn committed:
Merge branch 'master' into patch-1

eunrcn committed:
Merge pull request #2 from eunrcn/patch-1

Update README.md

eunrcn committed:
Final

eunrcn committed:
Update README.md

eunrcn committed:
User Guide

eunrcn committed:
User Guide

eunrcn committed:
Fixed Exceptions

eunrcn committed:
trying

eunrcn committed:
Final

eunrcn committed:
Final amends

eunrcn committed:
I literally owe my life to Jawad
